data/src/main/java/com/linkedin/data/codec/entitystream/AbstractDataDecoder.java

```diff
@@ -0,0 +1,401 @@
+            return;
+          default:
+            handleException(new Exception("Unexpected token " + token + " from data parser"));
+        }
+      }
+    }
+    catch (IOException e)
+    {
+      handleException(e);
+    }
+  }
+
+  protected final void validate(NonBlockingDataParser.Token token)
+  {
+    if (!(token == NOT_AVAILABLE || _expectedTokens.contains(token)))
+    {
+      handleException(new Exception("Expecting " + _expectedTokens + " but got " + token));
+    }
+  }
+
+  private void push(DataComplex dataComplex, boolean isList)
+  {
+    if (_currObjSize >= 0)
+    {
+      _currObjSizeStack.push(_currObjSize);
+    }
+    if (_currObjSize == 0)
+    {
+      addValue(dataComplex);
+      return;
+    }
+
+    if (!(_isCurrList || _stack.isEmpty()))
+    {
+      _currFieldStack.push(_currField);
+      _currField = null;
+    }
+    _stack.push(dataComplex);
+    _isCurrList = isList;
+    updateExpected();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void pop()
+  {
+    // The stack should never be empty because of token validation.
+    assert !_stack.isEmpty() : "Trying to pop empty stack";
+
+    DataComplex tmp = _stack.pop();
+
+    if (tmp instanceof DataMapBuilder)
+    {
+      tmp = ((DataMapBuilder) tmp).convertToDataMap();
+    }
+
+    if (_stack.isEmpty())
+    {
+      _result = (T) tmp;
+      // No more tokens is expected.
+      _expectedTokens = NONE;
+    }
+    else
+    {
+      _isCurrList = _stack.peek() instanceof DataList;
+      if (_currObjSizeStack != null)
+      {
+        _currObjSize = _currObjSizeStack.pop();
+      }
+      if (!_isCurrList)
+      {
+        _currField = _currFieldStack.pop();
+      }
+      addValue(tmp);
+      updateExpected();
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  protected void addValue(Object value)
+  {
+    if (!_stack.isEmpty())
+    {
+      DataComplex currItem = _stack.peek();
+      int currItemSize = -1;
+      if (_isCurrList)
+      {
+        CheckedUtil.addWithoutChecking((DataList) currItem, value);
+        currItemSize = ((DataList) currItem).size();
+      }
+      else
+      {
+        if (currItem instanceof DataMapBuilder)
+        {
+          DataMapBuilder dataMapBuilder = (DataMapBuilder) currItem;
+          if (dataMapBuilder.smallHashMapThresholdReached())
+          {
+            _stack.pop();
+            DataMap dataMap = dataMapBuilder.convertToDataMap();
+            _stack.push(dataMap);
+            CheckedUtil.putWithoutChecking(dataMap, _currField, value);
+          }
+          else
+          {
+            dataMapBuilder.addKVPair(_currField, value);
+          }
+        }
+        else
+        {
+          CheckedUtil.putWithoutChecking((DataMap) currItem, _currField, value);
+          currItemSize = ((DataMap) currItem).size();
+        }
+        _currField = null;
+      }
+      if (_currObjSize != -1 && currItemSize != -1 && _currObjSize == currItemSize)
+      {
+        pop();
+      }
```

Comment:
This seems odd, why are you popping based on currObjSize? pop() should only be called when END_* tokens are processed.

Reply:
this is a gray area call which I took to have it here. Look at reply on @karthikrg comment on pop().
Lets discuss offline, I can move this to protobuf parser if extra memory space usage is fine

Reply:
My bad, I forgot protobuf codec doesn't have start/end tokens. So ,the current logic tries to do two different types of parsing: one based on start/end tokens and one based on ordinal and size. There is no benefit in mixing them, it would just make the code complicated. 

I think you can implement this by having an abstract decoder that doesn't have any special logic and then extend it to have DataMapBuilder optimizations for start/end token decoders and size based optimizations for ordinal/size decoders. 

You'd have to make push/pop/addValue overridable. And also have abstract methods to get Map/List container objects.

Reply:
Sure, so I will take the 2nd approach from other comment and use abstract methods for initializing dataComplex. won't need to expose push/pop methods